### PR TITLE
Test for proxy location change

### PR DESCRIPTION
**Pull request by yschimke**
Demonstrate a proxy location changing via DNS.

```
Request to Proxy1

DNS Switch
if clean shutdown, close the Proxy1 socket.
Request to Proxy2

Request to Proxy2
```

### CODE REVIEW

**Comment by yschimke** [comment]
Without Clean shutdown.  

First and third requests succeed. The middle request fails on Proxy1 pooled connection.

```
Proxy Server 1 is localhost/127.0.0.1:51350
Proxy Server 2 is localhost/127.0.0.1:51351


Request to localhost/127.0.0.1:51350
ConnectStart(timestampNs=13381948498875, route=localhost:51350 via proxy 127.0.0.2:8008, call=okhttp3.internal.connection.RealCall@38239ea6)
Socket connection to: localhost/127.0.0.1:51350 was: proxyServer1/127.0.0.2:8008
ConnectEnd(timestampNs=13381959232750, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer1/127.0.0.2:8008 cipherSuite=none protocol=http/1.1}, route=localhost:51350 via proxy 127.0.0.2:8008, call=okhttp3.internal.connection.RealCall@38239ea6)
ConnectionAcquired(timestampNs=13381960056000, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer1/127.0.0.2:8008 cipherSuite=none protocol=http/1.1}, call=okhttp3.internal.connection.RealCall@38239ea6)
ConnectionReleased(timestampNs=13381968663208, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer1/127.0.0.2:8008 cipherSuite=none protocol=http/1.1}, call=okhttp3.internal.connection.RealCall@38239ea6)
server1.requestCount 1


Request to localhost/127.0.0.1:51351
ConnectionAcquired(timestampNs=13381971283291, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer1/127.0.0.2:8008 cipherSuite=none protocol=http/1.1}, call=okhttp3.internal.connection.RealCall@3163987e)
NoNewExchanges(timestampNs=13391981866666, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer1/127.0.0.2:8008 cipherSuite=none protocol=http/1.1})
ConnectionReleased(timestampNs=13391983153583, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer1/127.0.0.2:8008 cipherSuite=none protocol=http/1.1}, call=okhttp3.internal.connection.RealCall@3163987e)
ConnectionClosed(timestampNs=13391983677500, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer1/127.0.0.2:8008 cipherSuite=none protocol=http/1.1})


Request to localhost/127.0.0.1:51351
ConnectStart(timestampNs=13391985595000, route=localhost:51350 via proxy 127.0.0.3:8008, call=okhttp3.internal.connection.RealCall@7ca594c5)
Socket connection to: localhost/127.0.0.1:51351 was: proxyServer2/127.0.0.3:8008
ConnectEnd(timestampNs=13391986953208, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer2/127.0.0.3:8008 cipherSuite=none protocol=http/1.1}, route=localhost:51350 via proxy 127.0.0.3:8008, call=okhttp3.internal.connection.RealCall@7ca594c5)
ConnectionAcquired(timestampNs=13391987753833, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer2/127.0.0.3:8008 cipherSuite=none protocol=http/1.1}, call=okhttp3.internal.connection.RealCall@7ca594c5)
ConnectionReleased(timestampNs=13391989002833, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer2/127.0.0.3:8008 cipherSuite=none protocol=http/1.1}, call=okhttp3.internal.connection.RealCall@7ca594c5)
server2.requestCount 1
ConnectionClosed(timestampNs=13391992196375, connection=Connection{localhost:51350, proxy=HTTP @ myproxy:8008 hostAddress=proxyServer2/127.0.0.3:8008 cipherSuite=none protocol=http/1.1})
```

File: okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt [hunk 1]
```diff
@@ -102,6 +102,8 @@ open class RecordingConnectionListener(
   }
 
   private fun logEvent(e: ConnectionEvent) {
+    println(e)
```
Comment (swankjesse) [inline okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt:105]: Platform.log() ??

File: okhttp-testing-support/src/main/kotlin/okhttp3/SpecificHostSocketFactory.kt [hunk 2]
```diff
@@ -41,6 +41,7 @@ class SpecificHostSocketFactory(
       override fun connect(endpoint: SocketAddress?, timeout: Int) {
         val requested = (endpoint as InetSocketAddress)
         val inetSocketAddress = hostMapping[requested.address] ?: defaultAddress ?: requested
+        println("Socket connection to: ${inetSocketAddress} was: ${requested}")
```
Comment (swankjesse) [inline okhttp-testing-support/src/main/kotlin/okhttp3/SpecificHostSocketFactory.kt:44]: Platform.log() ??

**Review by swankjesse — APPROVED** [review]
I’d prefer no printlns but the test looks good.

File: okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt (hunk 1 above)
Comment (yschimke) [reply okhttp-testing-support/src/main/kotlin/okhttp3/RecordingConnectionListener.kt:105]: Think I'll remove

**Pull request merged**